### PR TITLE
SDR replace Reshape class with method.

### PR DESCRIPTION
**Pull request by ctrl-z-9000-times**
As discussed in issue #482.

### CODE REVIEW

File: src/nupic/types/Sdr.cpp [hunk 1]
```diff
@@ -135,6 +135,24 @@ namespace sdr {
     SparseDistributedRepresentation::~SparseDistributedRepresentation()
         { deconstruct(); }
 
+
+    void SparseDistributedRepresentation::reshape(const vector<UInt> &dimensions) const {
+        // Make sure we have the data in a format which does not care about the
+        // dimensions, IE: dense or sparse but not coordinates
+        if( not dense_valid and not sparse_valid )
```
Comment (breznak) [inline src/nupic/types/Sdr.cpp]: why is both dense & sparse required here? getSparse() sets only sparse_valid

File: src/nupic/types/Sdr.cpp [hunk 2]
```diff
@@ -135,6 +135,24 @@ namespace sdr {
...
+            getSparse();
+        coordinates_valid = false;
+        coordinates_.assign( dimensions.size(), {} );
+        dimensions_ = dimensions;
+        const auto oldSize = size;
+        // Calculate the SDR's size.
+        UInt newSize = 1;
+        for(UInt dim : dimensions)
+            newSize *= dim;
```
Comment (breznak) [inline src/nupic/types/Sdr.cpp]: nit, please consider replacing this chunk of code with a one liner: 
` int product = std::accumulate(v.begin(), v.end(), 1, std::multiplies<int>());`

File: src/nupic/types/Sdr.cpp [hunk 3]
```diff
@@ -237,29 +255,9 @@ namespace sdr {
...
-        // Subclasses may override these getters and ignore the valid flags...
-        if( !dense_valid and !sparse_valid and !coordinates_valid ) {
-            const auto data = value.getSparse();
-            sparse_.assign( data.begin(), data.end() );
-            sparse_valid = true;
-        }
-        do_callbacks();
+        NTA_CHECK( value.dimensions == dimensions )
+                    << "Failed to assign value=" << value << " to SDR=" << *this;
+        setSparse( (const SDR_sparse_t) value.getSparse() );
```
Comment (breznak) [inline src/nupic/types/Sdr.cpp]: nit, if you do the changes, use static_cast<SDR_sparse_t>(xx)

File: src/nupic/types/Sdr.hpp [hunk 4]
```diff
@@ -239,6 +239,11 @@ class SparseDistributedRepresentation : public Serializable
      */
     const UInt &size = size_;
 
+    /**
+     * Change the dimensions of the SDR.  The total size must not change.
+     */
+    void reshape(const std::vector<UInt> &dimensions) const;
```
Comment (breznak) [inline src/nupic/types/Sdr.hpp]: consider also returning the SDR? (*this), to allow chaining?

File: src/nupic/types/Sdr.cpp (hunk 1 above)
Comment (ctrl-z-9000-times) [reply src/nupic/types/Sdr.cpp]: Either dense or sparse is required here.
If neither dense nor sparse is available, then it gets the sparse data.

After it has gotten the data in a format which can be reshaped (sparse or dense) then it invalidates the coordinate format which can not be reshaped.

File: src/nupic/types/Sdr.hpp (hunk 4 above)
Comment (ctrl-z-9000-times) [reply src/nupic/types/Sdr.hpp]: I'd rather do this in a separate PR, and change all of the C++ methods at once.  I will open an issue for this.

Comment (ctrl-z-9000-times) [reply src/nupic/types/Sdr.hpp]: See issue #487

**Comment by ctrl-z-9000-times** [comment]
Breznak,
I changed the things which you wanted changed.  I liked the `accumulate` one-liner!  This is ready for another round of review when you get the chance.
Thanks

**Pull request merged**